File: Inaugura.Dialogic/DialogicHardware.cpp
// This is the main DLL file.

#include "stdafx.h"
#include "DialogicHardware.h"
#include "DialogicLine.h"

#using <System.DLL>
#using <System.ServiceProcess.DLL>

using namespace System;
using namespace System::Threading;
using namespace System::Collections;
using namespace System::ComponentModel;
using namespace System::ServiceProcess;

namespace Inaugura
{
namespace Dialogic
{	
	DialogicLine* DialogicHardware::Start() []
    {
		int numberOfBoards = DialogicHardware::GetBoardCount();			
		Console::WriteLine("Number of Bards Detected: {0}",numberOfBoards.ToString()); 
								
		String __gc* channelNames = DialogicHardware::GetChannelNames();				
		Console::WriteLine("Lines Detected: {0}",channelNames); 

		String* channels[] = DialogicHardware::ParseChannelNames(channelNames);
		DialogicLine* lines[] = new DialogicLine* [channels->Length];

		for(int i = 0; i < channels->Length; i++)
		{
			DialogicLine __gc* l = new DialogicLine(channels[i]);
			lines[i] = l;
		}		

		return lines;    
    }

	void DialogicHardware::StartDialogicService()
	{
		ServiceController __gc *dialogicService = new ServiceController("Dialogic");		
		if(dialogicService->Status != ServiceControllerStatus::Running)
		{
			Console::WriteLine("Starting {0}",dialogicService->DisplayName);
			dialogicService->Start();
			while(dialogicService->Status != ServiceControllerStatus::Running)
			{
				Console::Write(".");
				Thread::Sleep(1000);
				dialogicService->Refresh();
			}					
			Console::WriteLine("");
		}
	}

	void DialogicHardware::StopDialogicService()
	{
		ServiceController __gc *dialogicService = new ServiceController("Dialogic");		
		if(dialogicService->Status == ServiceControllerStatus::Running)
		{
			Console::WriteLine("Stoping {0}",dialogicService->DisplayName);
			dialogicService->Stop();
			while(dialogicService->Status == ServiceControllerStatus::Running)
			{
				Console::Write(".");
				Thread::Sleep(1000);
				dialogicService->Refresh();
			}					
			Console::WriteLine("");
		}	
	}

	void DialogicHardware::Stop()
    {
		this->StopDialogicService();		
    }

	bool DialogicHardware::IsDialogicServiceStarted()
	{
		ServiceController __gc *dialogicService = new ServiceController("Dialogic");		
			if(dialogicService->Status == ServiceControllerStatus::Running)
				return true;

			return false;
	}

	String* DialogicHardware::ParseChannelNames(String __gc* channelNames)[]
	{
		ArrayList __gc* a = new ArrayList();

			while(true)
			{
				int index = channelNames->IndexOf(";");
				if(index == -1)
					break;
				
				String __gc* channel = channelNames->Substring(0,index);
				a->Add(channel);				
				channelNames = channelNames->Substring(index+1);
			}

			String* channels[] = new String*[a->Count];
			for(int i = 0; i < channels->Length; i++)
			{
				channels[i]=a->get_Item(i)->ToString();
			}

			return channels;
	}

	int DialogicHardware::GetBoardCount()
	{
		try
		{
			return DialogicLegacy::GetBoardCount();
		}
		catch(char* str)
		{
			ApplicationException __gc *ex = new ApplicationException(str);
			throw ex;			
		}
	}

	String __gc* DialogicHardware::GetChannelNames()
	{
		
		try
		{
			String __gc* str = DialogicLegacy::GetChannelNames();
			return str;
		}
		catch(const char* str)
		{
			ApplicationException __gc *ex = new ApplicationException(str);
			throw ex;
		}
		return "";
	}	
}
}

File: Inaugura.Telephony.Dialogic/DialogicHardware.cs
using System;
using System.Xml;

using Inaugura;
using Inaugura.Dialogic;
using Inaugura.Telephony;

namespace Inaugura.Telephony.Dialogic
{
	/// <summary>
	/// 
	/// </summary>
	public class DialogicHardware : Inaugura.Telephony.TelephonyHardware
	{
		#region Variables
		private Inaugura.Dialogic.DialogicHardware mHardware;
		#endregion

		#region Properties
		public override string Name
		{
			get
			{
				return "Dialogic Hardware (VoiceAPI)";
			}
		}
		#endregion

		public override TelephonyLine[] Start()
		{
            mHardware = new Inaugura.Dialogic.DialogicHardware();
            DialogicLine[] lines = null;
			Inaugura.Dialogic.DialogicLine[] dialogicLines = null;

			try
			{
				if(!Inaugura.Dialogic.DialogicHardware.IsDialogicServiceStarted())
				{
					Log.AddLog("Starting Dialogic Service");
					Inaugura.Dialogic.DialogicHardware.StartDialogicService();
					Log.AddLog("Dialogic Service Started");
				}

				int numberOfBoards = Inaugura.Dialogic.DialogicHardware.GetBoardCount();
				Log.AddLog("Number of Boards: "+numberOfBoards.ToString());
			
				string channelNames = Inaugura.Dialogic.DialogicHardware.GetChannelNames();
				
				//string[] channels = Inaugura.Dialogic.DialogicHardware.ParseChannelNames(channelNames);
		
				/*
				dialogicLines = new Inaugura.Dialogic.DialogicLine[channels.Length];
				for(int i = 0; i < channels.Length; i++)
				{
					Log.AddLog("Creating Line: "+channels[i]);
					Inaugura.Dialogic.DialogicLine l = new Inaugura.Dialogic.DialogicLine(channels[i]);
					dialogicLines[i] = l;
				}
				*/		
				
				string[] channels = Inaugura.Dialogic.DialogicHardware.ParseChannelNames(channelNames);
		
				dialogicLines = new Inaugura.Dialogic.DialogicLine[channels.Length];
				for(int i = 0; i < channels.Length;i++)//channels.Length; i++)
				{
					Log.AddLog("Creating Line: "+channels[i]);
					Inaugura.Dialogic.DialogicLine l = new Inaugura.Dialogic.DialogicLine(channels[i]);
					dialogicLines[i] = l;
				}
						



				lines = new DialogicLine[dialogicLines.Length];
				for(int i = 0; i < dialogicLines.Length; i++)
				{	
					lines[i] = new DialogicLine(this,dialogicLines[i]);

					// TESTING
					//lines[i] = new DialogicFakeCallerIdLine(dialogicLines[i]);
				}
			}
			catch(Exception ex)
			{
				Log.AddLog(ex.Message);
				throw ex;
			}
			return lines;    			
		}	

		public override void Stop()
		{	
			/*
			if(Inaugura.Dialogic.DialogicHardware.IsDialogicServiceStarted())
			{
				Log.AddLog("Stopping Dialogic Service");
				Inaugura.Dialogic.DialogicHardware.StopDialogicService();
				Log.AddLog("Dialogic Service Stopped");
			}
			*/
		}

		public DialogicHardware()
		{
			//this.FileReferences.Add(new FileReference("OrbisTel.Telephony.Dialogic.dll", "TelephonyHardware", System.Reflection.Assembly.GetExecutingAssembly().GetName().Version.ToString()));
			//this.FileReferences.Add(new FileReference("Inaugura.Dialogic.dll", "TelephonyHardware", System.Reflection.Assembly.GetExecutingAssembly().GetName().Version.ToString()));
		}

		public DialogicHardware(XmlNode node) : base(node)
		{
		}

		public override string ToString()
		{
			return "Dialogic Hardware";
		}
	}
}
